src/pdf.sandbox.js

```diff
@@ -24,86 +25,124 @@ const TESTING =
   typeof PDFJSDev === "undefined" || PDFJSDev.test("!PRODUCTION || TESTING");
 
 class Sandbox {
-  constructor(module) {
-    this._evalInSandbox = module.cwrap("evalInSandbox", null, [
-      "string",
-      "int",
-    ]);
-    this._dispatchEventName = null;
+  constructor(module, externalNames) {
+    this._commFun = module.cwrap("commFun", null, ["string", "string"]);
     this._module = module;
-    this._alertOnError = 1;
+    this._alertOnError = 0;
+    this._externalNames = externalNames;
   }
 
   create(data) {
-    const sandboxData = JSON.stringify(data);
-    const extra = [
-      "send",
-      "setTimeout",
-      "clearTimeout",
-      "setInterval",
-      "clearInterval",
-      "crackURL",
-    ];
-    const extraStr = extra.join(",");
-    let code = [
-      "exports = Object.create(null);",
-      "module = Object.create(null);",
+    const sandboxData = Sandbox.cloneIn(data);
+    // In the sandox, external functions like send, setTimeout, ... exist
+    // as global variables.
+    const external = `{ ${this._externalNames.join(",")} }`;
```

Comment:
So in order to define an external X, you currently have to do:

1. add implementation named X to pdf.sandbox.external.js
2. declare X in `myjs.js`, twice: first the global, then again in `$externals` - https://github.com/mozilla/pdf.js.quickjs/blob/895a07683ae7dc71484ce59f493a5c2e52900a7b/src/myjs.js#L23
3. declare C function with name X for quickjs in quickjs_extra.c - https://github.com/mozilla/pdf.js.quickjs/blob/895a07683ae7dc71484ce59f493a5c2e52900a7b/src/quickjs_extra.c
4. reference externals.X from the actual JS implementation.

2 and 3 is an unfortunate amount of repetition, each with potential for bugs. Step 3 already has bugs: You're returning `JS_EXCEPTION` several times without setting an exception first, i.e. by `JS_ThrowTypeError(ctx, "message here");`. The effect of this is that the function won't actually throw an exception.

2 and 3 can be reduced to a zero-maintenance overhead + behavior equivalent to m-c, by adding a way to make generic functions from inside the sandbox to outside the sandbox, along with a good choice of a serialization and deserialization function. I suggest JSON serialization because a serializer is already available in quickjs's C code, and the corresponding deserializer is available in normal JS. Without JSON serialization, you would need considerably more code (like you have right now...) for no apparent benefit.

Here is an example where you'll maintain a table of callable methods in the `pdf.sandbox.js`, and allow the sandboxed code to call any of the methods in the table. This is not the only way to implement this concept, but I want to show at least one example for clarification.

```c
static JSValue js_callExternalFunc(JSContext *ctx, JSValueConst this_val,
                                int argc, JSValueConst *argv) {
  // TODO: function name (char*) from argv
  // TODO: rest of arguments OR just argv[1] as array, JS_JSONStringify it.
  // TODO: call external function with function name + args
  // TODO: Consider optional exception handling.
}
```

In myjs.js, you'll only need one export - that generic external function handler.

In pdf.sandbox.js, make that generic external function available. ensure that name is valid, parse args from JSON and call `externals[name](...args)`.

What I described above is the rudimentary version of how m-c's sandbox handles function calls of exported functions. The difference is that in my implementation (which matches how you're currently using it) is that values are copied rather than passed by a (mutable) reference. Since you're only passing numbers and strings, this distinction doesn't matter at all.

With the above logic, you could do something like 

`callExternalFunction("alert", ["msg"])` or
`callExternalFunction("alert", "msg")`, depending on your signature.

or specifically in this context:

```
var externals = {};
for (let ext of ${externals}) {
  externals[ext] = callExternalFunction.bind(null, ext);
}
delete callExternalFunction; // TODO: does this work? otherwise = null.
```

This can probably be made prettier, but at least it works with less maintenance cost than what you have right now.